code/go/0chain.net/chaincore/block/entity.go

```diff
@@ -458,13 +458,15 @@ func (b *Block) HashBlock() {
 }
 
 /*ComputeTxnMap - organize the transactions into a hashmap for check if the txn exists*/
-func (b *Block) ComputeTxnMap() {
+func (b *Block) StartComputeTxnMap() {
 	b.mutexTxns.Lock()
-	defer b.mutexTxns.Unlock()
-	b.TxnsMap = make(map[string]bool, len(b.Txns))
-	for _, txn := range b.Txns {
-		b.TxnsMap[txn.Hash] = true
-	}
+	go func() {
+		defer b.mutexTxns.Unlock()
+		b.TxnsMap = make(map[string]bool, len(b.Txns))
+		for _, txn := range b.Txns {
+			b.TxnsMap[txn.Hash] = true
+		}
+	}()
```

Comment:
Do not use the `b.mutextTxns`  in different gorountines. This could cause deadlock, when the mutex is locked, but before the mutex is unlocked in the goroutine, another thread could acquire the `RLock`, before it calls the `RUnlock`, if the goroutine ended and try to call the `deter b.mutexTxns.Unlock()`, we will get deadlock.

Reply:
I don't think such a deadlock is possible, since no goroutine will be allowed to aquire lock after ```b.mutexTxns.Lock()``` is called. Goroutine with ```defer b.mutexTxns.Unlock()``` will be guaranteed to execute eventually, so this lock will be unlocked. @peterlimg is right, this code has a problem: we should try make locks as short as possible (this mutext is more like monitor for TxnMap) but in the code above we can't guarantee than goroutine will be started immediately, so this lock will be locked extra time which is not necessary

Reply:
Right, in Golang (unlike some other programming languages) a locked RWMutex is not associated with a particular goroutine. One goroutine may RLock (Lock) a RWMutex and then arrange for another goroutine to RUnlock (Unlock) it.
My main concern was that upon initialization of a new block (in `protocol_block_main.go` function `GenerateBlock` we fill the block with transactions `b.Txns = append(b.Txns, txn)` and then start map calculation in a goroutine.
As the goroutine start may be delayed, `Block.HasTransaction()` and `Block.Clone()` may incorrectly handle the empty map.
In this patch, they have to wait until map calculation has finished and that seems logical to me.
How would you suggest to handle this situation?

Reply:
Oh, yes, the deadlock would not exist here, my mistake. Basically, once you acquired the write lock, all others has to wait for the lock to be unlocked, so we don't need to put the `b.TxnsMap` updating code in a new goroutine.